### PR TITLE
user pre-built storage image

### LINKED ISSUE #124: Make integration tests faster
We currently build containers for each push, but perhaps we should separate out those steps and push containers to GHCR (specifically the SPDK container) to make things faster. Worth discussing on this issue how we want to do this.

now they take 20m

image

### PR DESCRIPTION
**Pull request by glimchb**
Fixes #124

Signed-off-by: Boris Glimcher <Boris.Glimcher@emc.com>

### CODE REVIEW

**Comment by glimchb** [comment]
why still building SPDK ? see https://github.com/opiproject/opi-poc/runs/7436446179?check_suite_focus=true
```
Building spdk-target
Sending build context to Docker daemon  9.728kB
Step 1/12 : FROM fedora:36 as build
 ---> 98ffdbffd207
Step 2/12 : ARG TAG=v22.05
 ---> Running in 52cdf5e0be5e
Removing intermediate container 52cdf5e0be5e
 ---> 88b4dcd10ed8
Step 3/12 : ARG ARCH=native
 ---> Running in 310c8e3ce234
Removing intermediate container 310c8e3ce234
 ---> 018bc8e2f237
Step 4/12 : WORKDIR /root
 ---> Running in 069026cfb5f9
Removing intermediate container 069026cfb5f9
 ---> cb9f11bd5124
Step 5/12 : RUN dnf install -y git rpm-build diffutils procps-ng && dnf clean all
...
```

**Comment by seroyer** [comment]
My guess is that we aren't uploading/downloading the image from a known repo.

> Successfully tagged ghcr.io/opiproject/opi-spdk:main
> Image for service spdk-target was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`

To be fair, I'm not super familiar with ghcr or docker compose, but normally there'd need to be a docker push to push it to the container registry after building it, and a docker pull to (or a run/FROM) to cause it to pull from the container registry.

**Comment by glimchb** [comment]
> My guess is that we aren't uploading/downloading the image from a known repo.
> 
> > Successfully tagged ghcr.io/opiproject/opi-spdk:main
> > Image for service spdk-target was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`
> 
> To be fair, I'm not super familiar with ghcr or docker compose, but normally there'd need to be a docker push to push it to the container registry after building it, and a docker pull to (or a run/FROM) to cause it to pull from the container registry.

it was already pushed, see https://github.com/opiproject/opi-poc/pkgs/container/opi-spdk
```
docker pull ghcr.io/opiproject/opi-spdk:main
```

**Comment by seroyer** [comment]
And on another point, it's not trivial to cache container images.  As @mestery is fond of pointing out, cache invalidation is a hard problem.  How do we know if we need to build an updated image vs using an existing one from a container registry?  It's much simpler to either build every time, or separate that logic into two separate jobs:
1) an spdk job that runs when spdk updates
2) an opi job that just consumes the latest (never builds it)

**Comment by glimchb** [comment]
> And on another point, it's not trivial to cache container images. As @mestery is fond of pointing out, cache invalidation is a hard problem. How do we know if we need to build an updated image vs using an existing one from a container registry? It's much simpler to either build every time, or separate that logic into two separate jobs:
> 
> 1. an spdk job that runs when spdk updates
> 2. an opi job that just consumes the latest (never builds it)

the split already happened. I build and push SPDK image in `storage` CI and consume it in `integration` CI

**Review by mestery — APPROVED** [review]
This seems reasonable.

File: integration/docker-compose.spdk.yml [hunk 1]
```diff
@@ -3,6 +3,7 @@ version: "3.7"
 services:
 
   spdk-target:
+    image: ghcr.io/opiproject/opi-spdk:main
     build:
```
Comment (seroyer) [inline integration/docker-compose.spdk.yml:7]: If we never want to build it from here, should we remove the build context?

**Comment by glimchb** [comment]
Looks like our package is private here https://github.com/orgs/opiproject/packages/container/opi-spdk/settings
```
Change package visibility
This package is currently private.
```
I can't change this to public for some reason @mestery @seroyer check if you can please

File: integration/docker-compose.spdk.yml (hunk 1 above)
Comment (glimchb) [reply integration/docker-compose.spdk.yml:7]: left context here for the case people can't access ghcr or want to change dockerfile locally

**Comment by glimchb** [comment]
Made a package public, let's see now...

**Pull request merged**